### PR TITLE
fix parseLabel to be proto3 and proto3_optional aware

### PR DESCRIPTION
**Pull request by jhump**
This also updates the handling of oneofs to ignore synthetic oneofs (a synthetic oneof is generated on behalf of a proto3_optional field).

### CODE REVIEW

File: wire-library/wire-protoc/src/main/java/com/squareup/wire/protocwire/WireGenerator.kt [hunk 1]
```diff
@@ -151,7 +151,7 @@ private fun parseFileDescriptor(fileDescriptor: FileDescriptorProto, descs: Desc
   val types = mutableListOf<TypeElement>()
   val baseSourceInfo = SourceInfo(fileDescriptor, descs)
   for ((sourceInfo, messageType) in fileDescriptor.messageTypeList.withSourceInfo(baseSourceInfo, FileDescriptorProto.MESSAGE_TYPE_FIELD_NUMBER)) {
-    types.add(parseMessage(sourceInfo, packagePrefix, messageType))
+    types.add(parseMessage(sourceInfo, packagePrefix, messageType, (fileDescriptor.syntax ?: "") == "proto3"))
```
Comment (buildbreaker) [inline wire-library/wire-protoc/src/main/java/com/squareup/wire/protocwire/WireGenerator.kt]: Wire does have an enum for the Syntax.

I use it here on L173:
```
syntax = if (fileDescriptor.hasSyntax()) Syntax[fileDescriptor.syntax] else Syntax.PROTO_2,
```

Maybe we can pass the enum through instead?

File: wire-library/wire-protoc/src/main/java/com/squareup/wire/protocwire/WireGenerator.kt [hunk 2]
```diff
@@ -312,7 +322,7 @@ private fun indexFieldsByOneOf(
 ): Map<Int, List<FieldElement>> {
   val oneOfMap = mutableMapOf<Int, MutableList<FieldElement>>()
   for ((descriptor, fieldElement) in fields) {
-    if (descriptor.hasOneofIndex()) {
+    if (descriptor.hasOneofIndex() && !descriptor.proto3Optional) {
```
Comment (buildbreaker) [inline wire-library/wire-protoc/src/main/java/com/squareup/wire/protocwire/WireGenerator.kt:326]: Ah. That's what you meant in the channel thread.

**Review by buildbreaker — APPROVED** [review]
Looks good but can we use the enum for the syntax?

File: wire-library/wire-protoc/src/main/java/com/squareup/wire/protocwire/WireGenerator.kt (hunk 1 above)
Comment (jhump) [reply wire-library/wire-protoc/src/main/java/com/squareup/wire/protocwire/WireGenerator.kt]: Done

**Pull request merged**